Updated Spreedly::Environment for Google Pay

guillec opened the pull request:
# Objective

This PR adds the ability for Spreedly to handle GooglePay transactions.

# How

The idea is that users will now be able to pass a `payment_method` key in the options argument of the `purchase_on_gateway`  method. This key will let the Spreedly gem know if we are sending a GooglePay token. 
In the future we can pass apple_pay as well.

Example:
```
purchase_on_gateway(gateway_token, payment_method_token, amount, options = {"payment_method" => "google_pay"})
```

Here is an example of what Spreedly requires for Google:

https://docs.spreedly.com/guides/google-pay/#transactions

```
<transaction>
      <amount>1122</amount>
      <currency_code>USD</currency_code>
      <google_pay>
        <payment_data>
          <![CDATA[{"signature":"MEUCIA6SGVRwhOyeYRkeDUUNwB/kGtyfQAlOsg7NZydT17u/AiEA48BhWGQEF1EbEU0J+m8eSK3rTfhok9QqpiFVbME+Ky0\u003d","protocolVersion":"ECv1","signedMessage":"{\"encryptedMessage\":\"3v4IcT/eovIDP2WF8iRUy4qWQnE9Cx0vQxIZ5f9i3Emv3Tzs1AzvB7cxXhxrjp9FVIzdOwsZAPAsm03gvoYq8Xdr70XvrVRd2MFwQhMC7IV/uEsthw4JsR8oCkbI5v/zqhu2B+JodFgavNliHcpKBgijy2D6bpx7jXEkM39M/L4oBObFxFrhVSLA1GjOV6A5gLAXNXt0ffkCYekihqAyJrWlk3sCBDCF5SUiAKEIOIZtzZLgusxjVp6ufZHOHm/53uhAi6JWSJ1E6G5aaYGtULYdwgURHtxN5OIzQPYlEGctaQd5tgfCsBFfGkYyN1GRNgclbaLzAfk/Jn7/6IVKuV0ol3xubTcnjGTZXwtTjiEyYDoz1yVqB9ViMmJa55L6nBtbbAkcNEgAi7dPnrbvBGEP7QWsNT9D71g8SWrlRTCYUAOyuamaQhofG4ul1IVjmltdAy2BHBWpqgJnR9kczydQyE7uDiqhSC1/0eG8GCGIqoi8XfOioGXfMyLZ1p2ZcNK9ECjzUrH/edrwgtShxgWuWMwQTM4DQlVTAA/R4DVs192YWZcc7jm5wLqZ0+XEaPuighJM1Ps1Egeccg\\u003d\\u003d\",\"ephemeralPublicKey\":\"BA2SvF9BdCX7Tl1wwRkyLzTfqhctobhZgSugC9Cz77XNUCBOBMfFyJQt506PUs89D6IJZZfOkZopy0shRF9Uph4\\u003d\",\"tag\":\"Uhin1BE7KAuuiam7eEQFimRUDd9Xn6tZc2fClTpfrXQ\\u003d\"}"}
]]>
        </payment_data>
        <test_card_number>4111111111111111</test_card_number> <!-- Remove in production -->
      </google_pay>
    </transaction>
```

guillec committed:
Updated Spreedly::Environment for Google Pay

This commit updates the logic in auth_purchase_body to handle payment
tokens related to google pay.

It does this by looking at the payment token string and if it has the
word google_pay we know it is a google payment token.

guillec commented:
@hoenth @kcpavan this is the first part of what I believe is what we need todo for spreedly. 
Would like to get your thoughts on this approach.

kcpavan reviewed (commented):
We receive string similar to this from google pay 
`{
  \"id\": \"tok_1MifVB2eZvKYlo2Cr1MdnwcK\",
  \"object\": \"token\",
  \"card\": {
    \"id\": \"card_1MifVB2eZvKYlo2CHsV0mjPM\",
    \"object\": \"card\",
    \"address_city\": \"Mountain View\",
    \"address_country\": \"US\",
    \"address_line1\": \"1600 Amphitheatre Parkway\",
    \"address_line1_check\": \"unchecked\",
    \"address_line2\": null,
    \"address_state\": \"CA\",
    \"address_zip\": \"94043\",
    \"address_zip_check\": \"unchecked\",
    \"brand\": \"Visa\",
    \"country\": \"US\",
    \"cvc_check\": null,
    \"dynamic_last4\": \"1111\",
    \"exp_month\": 12,
    \"exp_year\": 2025,
    \"funding\": \"credit\",
    \"last4\": \"1111\",
    \"metadata\": {},
    \"name\": null,
    \"tokenization_method\": \"android_pay\"
  },
  \"client_ip\": \"209.85.198.210\",
  \"created\": 1678114741,
  \"livemode\": false,
  \"type\": \"card\",
  \"used\": false
}" ` 

Your mock data in spec looks different, are we making any additional api calls to get this xml data?

guillec commented:
Discussed with Pavan and we are getting back a different looking token than what the spreedly docs are showing. We decided to leave this here until we have the front end done and can have some better example of what the token will look like.

guillec committed:
Updated google pay token

The token I had in the stub was not correct. This commit adds the
corrected format.

guillec committed:
Updated method to handle google_pay token

Updated the add_payment_token to look into options for a payment_method
key. Users will be able to pass a payment_method to specify if this is
for google_pay or in the future it could be apple_pay.

guillec committed:
Updated spec to test a google pay token

guillec committed:
Handle spreedly response to GooglePay

this commit adds the logic and test for handling a google pay response
from spreedly.

guillec committed:
Remove attributes from GooglePay payment method

These attributes are not needed like in Paypal.

kcpavan commented on lib/spreedly/environment.rb:
```diff
@@ -174,14 +174,22 @@ def auth_purchase_body(amount, payment_method_token, options)
 
+    def add_payment_token(doc, payment_method_token, options = {})
+      if options["payment_method"] == "google_pay"
+        doc << "<google_pay><payment_data><![CDATA[#{payment_method_token}]]></payment_data><test_card_number>4111111111111111</test_card_number></google_pay>"
```
We should add `<test_card_number>` only for non production environments right?

kcpavan replied:
```suggestion
       doc << <<~XML
          <google_pay>
            <payment_data><![CDATA[#{payment_method_token}]]></payment_data>
            #{'<test_card_number>4111111111111111</test_card_number>' unless Rails.env.production?}
          </google_pay>
        XML"
```

pawan-kc committed:
Set test_card_number for google_pay purchase request only if environment is non production

kcpavan reviewed (commented):
it requires one small change.

kcpavan commented:
We have one more repo for spreedly gem https://github.com/GiveCorps/spreedly-gem/commits/master, I see 3 weeks back commit on that repo by @hoenth . Should we continue with that repo or should we merge that changes here ?.

I do not have write permission for this repo.

kcpavan commented on lib/spreedly/environment.rb:
```diff
@@ -174,14 +174,22 @@ def auth_purchase_body(amount, payment_method_token, options)
     end
 
+    def add_payment_token(doc, payment_method_token, options = {})
+      if options["payment_method"] == "google_pay"
```
```suggestion
      if options[:payment_method] == :google_pay
```

kcpavan commented on test/unit/purchase_test.rb:
```diff
@@ -153,6 +155,26 @@ def test_request_body_params
 
+  def test_request_body_params_for_google_pay
+    body = get_request_body(successful_purchase_response) do
+      @environment.purchase_on_gateway("TheGatewayToken", google_pay_token, 2001, all_possible_options.merge("payment_method" => "google_pay"))
```
```suggestion
      @environment.purchase_on_gateway("TheGatewayToken", google_pay_token, 2001, all_possible_options.merge(payment_method: :google_pay))
```

kcpavan reviewed (changes requested):
Added few suggestions.

pawan-kc committed:
Allow test_mode option to be configured from hosting app

hoenth reviewed (approved):
LGTM

hoenth commented on lib/spreedly/environment.rb:
```diff
@@ -174,14 +174,27 @@ def auth_purchase_body(amount, payment_method_token, options)
+        doc << <<~XML
+          <google_pay>
+            <payment_data><![CDATA[#{payment_method_token}]]></payment_data>
+            #{'<test_card_number>4111111111111111</test_card_number>' if options[:test_mode]}
```
@guillec Is `options[:test_mode]` always false for production, and true for development, beta, test?

kcpavan replied:
Yes Tom, Here is how we are setting. 
`def payment_profile_specific_fields
    options = {}
    return options unless payment_profile.digital_wallet?

    options[:payment_method] = :google_pay if payment_profile.google_pay?
    options[:payment_method] = :apple_pay if payment_profile.apple_pay?
    options[:test_mode] = !Rails.env.production? # This is used for adding test card nubmer for Digital wallets in gem
    options
  end`

pawan-kc committed:
bumps the version

guillec committed:
Merge branch 'master' into accept_google_pay_flow

kcpavan reviewed (approved):
LGTM
